### ISSUE TITLE
Order dependent issue populating documents that contain embedded discriminators and refPath

### ISSUE THREAD

**Issue by robertjustjones** [Automattic/mongoose#5858]
This is a bug in 4.13.5 and master (using node 7.4, mongodb 3.4).

For population of subdocuments with a discriminator and refPath, the leaf can populate with null depending on document order.

The following can be dropped into /test and run.
```js
/**
 * Test dependencies.
 */

var _ = require('lodash');
var start = require('./common');
var async = require('async');
var assert = require('power-assert');
var mongoose = start.mongoose;
var utils = require('../lib/utils');
var random = utils.random;
var Schema = mongoose.Schema;
var ObjectId = Schema.ObjectId;
var DocObjectId = mongoose.Types.ObjectId;

/**
 * Tests.
 */

describe('model: populate:', function() {
  var User;
  var Comment;
  var BlogPost;
  var posts;
  var users;

  before(function() {
    User = new Schema({
      name: String,
      email: String,
      gender: {type: String, enum: ['male', 'female'], default: 'male'},
      age: {type: Number, default: 21},
      blogposts: [{type: ObjectId, ref: 'RefBlogPost'}],
      followers: [{type: ObjectId, ref: 'RefUser'}]
    });

    /**
     * Comment subdocument schema.
     */

    Comment = new Schema({
      asers: [{type: ObjectId, ref: 'RefUser'}],
      _creator: {type: ObjectId, ref: 'RefUser'},
      content: String
    });

    /**
     * Blog post schema.
     */

    BlogPost = new Schema({
      _creator: {type: ObjectId, ref: 'RefUser'},
      title: String,
      comments: [Comment],
      fans: [{type: ObjectId, ref: 'RefUser'}]
    });

    posts = 'blogposts_' + random();
    users = 'users_' + random();

    mongoose.model('RefBlogPost', BlogPost);
    mongoose.model('RefUser', User);
    mongoose.model('RefAlternateUser', User);
  });

  describe('github issues', function() {
    var db;

    before(function() {
      db = start();
    });

    after(function(done) {
      db.close(done);
    });

    it('discriminator child of child schemas (robertjustjones)', function(done) {
      var options = { discriminatorKey: 'kind' };
      var activitySchema = new Schema({ title: { type: String } }, options);

      var dateActivitySchema = new Schema({
        postedBy: { type: Schema.Types.ObjectId, ref: 'robertjustjones_User', required: true }
      }, options);

      var eventActivitySchema = new Schema({ test: String }, options);

      var logSchema = new Schema({
        seq: Number,
        activity: { type: Schema.Types.ObjectId, refPath: 'kind', required: true },
        kind: String,
      }, options);

      var User = db.model('robertjustjones_User', { name: String });
      var Activity = db.model('robertjustjones_Activity', activitySchema);
      var DateActivity = Activity.discriminator('robertjustjones_Date', dateActivitySchema);
      var EventActivity = Activity.discriminator('robertjustjones_Event', eventActivitySchema);
      var Log = db.model('robertjustjones_Log', logSchema);

      User.create({ name: 'val' }, function(error, user) {
        assert.ifError(error);
        var eventActivity = {
          title: 'test2',
          test: 'test'
        };
        EventActivity.create(eventActivity, function(error, eventActivityModel) {
          assert.ifError(error);
          var dateActivity = { title: 'test', postedBy: user._id };
          DateActivity.create(dateActivity, function(error, dateActivityModel) {
            assert.ifError(error);
            var log1 = {
              seq: 1,
              activity: eventActivityModel._id,
              kind: 'robertjustjones_Event',
            };
            Log.create(log1, function(error) { 
              assert.ifError(error);
              var log2 = {
                seq: 2,
                activity: dateActivityModel._id,
                kind: 'robertjustjones_Date',
              };
              Log.create(log2, function(error) { 
                assert.ifError(error);
                test();
              });
            });
          });
        });
      });

      function test() {
        var pop = {
          path: 'activity',
          populate: { path: 'postedBy' }
        };

        debugger;
        
        Log.find({}).populate(pop).sort({seq:-1}).exec(function(error, results) {
          assert.ifError(error);
          assert.equal(results.length, 2);
          assert.equal(results[1].activity.kind, 'robertjustjones_Event' );
          assert.equal(results[0].activity.kind, 'robertjustjones_Date' );
          assert.equal(results[0].activity.postedBy && results[0].activity.postedBy.name, 'val');
          assert.equal(results[1].activity.postedBy, null);
          
          Log.find({}).populate(pop).sort({seq:1}).exec(function(error, results) {
            assert.ifError(error);
            assert.equal(results.length, 2);
            assert.equal(results[0].activity.kind, 'robertjustjones_Event' );
            assert.equal(results[1].activity.kind, 'robertjustjones_Date' );
            assert.equal(results[1].activity.postedBy && results[1].activity.postedBy.name, 'val');
            assert.equal(results[0].activity.postedBy, null);
            done();
          });
        });
      }
    });

  });
});
```

**Comment by robertjustjones** [comment]
I believe the issue is in `Query.prototype._find` involving `mongooseOptions` in the scope outside `cb()`.

**Comment by robertjustjones** [comment]
For the second query, by the time that `getModelsMapForPopulate(models, docs, options)` is reached, 
```
docs[0].kind === 'robertjustjones_Date' (correct) 
model.modelName === 'robertjustjones_Date' (correct)
options.model.modelName === 'robertjustjones_Event' (incorrect)
```
This leads to an incorrect `modelMap`.

**Comment by vkarpov15** [comment]
Thanks for the detailed repro script, will fix for 4.13.7 :+1:

**Comment by vkarpov15** [comment]
Actually, that's strange, I can't reproduce this issue anymore, your test case succeeds against master now. I'm pretty baffled, I'm certain it was failing 2 days ago. Can you confirm you're still having this issue?

**Comment by robertjustjones** [comment]
Yes, I still see in on 2ecf9385 from master.

**Comment by vkarpov15** [comment]
Thanks for your patience, this issue proved very tricky to solve because it proved to be a subtle race condition that I couldn't repro consistently. Ended up being a one line fix. Fix will be in 4.13.8 :+1:

**Comment by robertjustjones** [comment]
Wonderful!  My attempted fixes were in lib/query, basically trying to clone downstream of where you did... knew it wasn't elegant enough to be right. Bravo.
